### PR TITLE
Add org.endlessos.Key

### PR DESCRIPTION
**Pull request by starnight**
### Please confirm your submission meets all the criteria

- [x] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [x] My pull request follows the instructions at [App Submission][submission].
- [x] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)*
- [x] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [x] I am an upstream contributor to the project. If not, I contacted upstream developers about submitting their software to Flathub. **Link:**
- [x] I own the domain used in the application ID or the domain has a policy for delegating subdomains (e.g. GitHub, SourceForge).
- [x] Any additional patches or files have been submitted to the upstream projects concerned. *(If not, explain why.)*

[reqs]: https://docs.flathub.org/docs/for-app-authors/requirements
[maint]: https://docs.flathub.org/docs/for-app-authors/maintanance
[submission]: https://docs.flathub.org/docs/for-app-authors/submission

### CODE REVIEW

File: org.endlessos.Key.yaml [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    builddir: true
+    run-tests: true
+    config-opts:
+      - -Dkolibri_home_template_dir=/app/share/kolibri-home-template
+      - -Dprofile=default
+    sources:
+      - type: git
+        url: https://github.com/endlessm/endless-key-flatpak.git
+        tag: endlesskey_1.0.0
+        commit: 927b8022cc25facc11fe38ccd20ab2c04b49da1b
```
Comment (dbnicholson) [inline org.endlessos.Key.yaml:64]: I changed the versioning and tagging a bit, so I think this should be:

```suggestion
        tag: v0.1
        commit: a9781eee6379ae2196ca2c81a65ee20a0a37aafe
```

**Review by dbnicholson — CHANGES_REQUESTED** [review]

File: org.endlessos.Key.yaml [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+      - find ${KOLIBRI_MODULE_PATH} -type d -name 'test*' -not -path ${KOLIBRI_MODULE_PATH}/dist/django/test -exec rm -rf '{}' '+'
+      - find /app/lib/python3.10 /app/share/kolibri-home-template -name '*.js.map' -exec rm '{}' '+'
+
+  - name: kolibri-tools
+    buildsystem: simple
+    build-commands:
+      - install -m 755 -t ${FLATPAK_DEST}/bin kolibri-listcontent.py
+    sources:
+      - type: dir
+        path: tools
```
Comment (dylanmccall) [inline org.endlessos.Key.yaml:81]: I think `kolibri-listcontent.py` is still useful for Endless Key, but I am reminded _now_ is a good time to quit having this in the flatpak repo. Wouldn't really help to move it to our kolibri-gnome fork. Maybe we can propose it upstream as a new management command? I don't think we actually _did_ at any point, and it makes more sense now that the other content manifest stuff is in place. (Although it needs a rewrite to reuse code that is now actually upstream).

File: org.endlessos.Key.yaml [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    merge-dirs: lib/python
+    subdirectories: true
+    no-autodownload: true
+
+build-options:
+  env:
+    KOLIBRI_MODULE_PATH: /app/lib/python3.10/site-packages/kolibri
+  test-args:
+    - --socket=fallback-x11
+    - --socket=wayland
```
Comment (hfiguiere) [inline org.endlessos.Key.yaml:38]: this won't work on the CI.

File: org.endlessos.Key.yaml [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+  - modules/python3-markdown.json
+  - modules/python3-kolibri.json
+  - modules/python3-kolibri-pytz.json
+  - modules/python3-kolibri-app-desktop-xdg-plugin.json
+  - modules/python3-kolibri-desktop-auth-plugin.json
+  - modules/kolibri-home-template.json
+  - modules/python3-setproctitle.json
+  - modules/python3-virtualenv-api.json
+  - modules/kolibri-content-dir.json
+  - modules/kolibri-plugins-dir.json
```
Comment (hfiguiere) [inline org.endlessos.Key.yaml:51]: non-generated manifest that are separate like that are a pain for maintenance.

File: tools/kolibri-listcontent.py
Comment (hfiguiere) [inline tools/kolibri-listcontent.py:1]: is there a reason to have this here and not upstrem?

File: org.endlessos.Key.yaml [hunk 5]
```diff
@@ -0,0 +1,81 @@
+app-id: org.endlessos.Key
+runtime: org.gnome.Platform
+runtime-version: '44'
+sdk: org.gnome.Sdk
+command: /app/bin/kolibri-gnome
```
Comment (hfiguiere) [inline org.endlessos.Key.yaml:5]:
```suggestion
command: kolibri-gnome
```

File: org.endlessos.Key.yaml [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+finish-args:
+  - --device=dri
+  - --share=ipc
+  - --share=network
+  - --socket=fallback-x11
+  - --socket=pulseaudio
+  - --socket=wayland
+  - --socket=x11
+  - --system-talk-name=org.endlessos.Key.Daemon
```
Comment (hfiguiere) [inline org.endlessos.Key.yaml:14]: system? How do you make sure it exists?

**Comment by hfiguiere** [comment]
bot, build org.endlessos.Key

File: org.endlessos.Key.yaml (hunk 3 above)
Comment (hfiguiere) [reply org.endlessos.Key.yaml:38]: As I was saying:

```
]2;flatpak-builder: Cleanup kolibri-gnomeError: Can't specify test-args in sandboxed build
```

File: org.endlessos.Key.yaml [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+  - modules/kolibri-home-template.json
+  - modules/python3-setproctitle.json
+  - modules/python3-virtualenv-api.json
+  - modules/kolibri-content-dir.json
+  - modules/kolibri-plugins-dir.json
+
+  - name: kolibri-gnome
+    buildsystem: meson
+    builddir: true
+    run-tests: true
```
Comment (hfiguiere) [inline org.endlessos.Key.yaml:56]:
```suggestion
```

File: org.endlessos.Key.yaml [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+        commit: 927b8022cc25facc11fe38ccd20ab2c04b49da1b
+
+  - name: python3-klobri-cleanup
+    buildsystem: simple
+    build-commands:
+      # GNOME 44 runtime has Python 3.10
+      - find ${KOLIBRI_MODULE_PATH}/dist/cext -type d -name 'cp*' -not -path ${KOLIBRI_MODULE_PATH}/dist/cext/cp310 -exec rm -rf '{}' '+'
+      - rm -rf ${KOLIBRI_MODULE_PATH}/dist/cext/*/Windows
+      - find ${KOLIBRI_MODULE_PATH} -type d -name 'test*' -not -path ${KOLIBRI_MODULE_PATH}/dist/django/test -exec rm -rf '{}' '+'
+      - find /app/lib/python3.10 /app/share/kolibri-home-template -name '*.js.map' -exec rm '{}' '+'
```
[the remark this answers is not in the document]
Comment (starnight) [reply org.endlessos.Key.yaml:73]: > This looks like like your own set of changes from https://github.com/flathub/org.learningequality.Kolibri/blob/master/org.learningequality.Kolibri.yaml#L68-L123. I'm really curious what the rationale is. Does this produce a similar result in terms of total size, or is there an issue with the original cleanup module?

Yes, I try to simplify the rm list.
I understand we are working on localization, but I do not know the locale part helps, or not.  So, I keep it right now.

File: org.endlessos.Key.yaml (hunk 2 above)
Comment (starnight) [reply org.endlessos.Key.yaml:81]: Yes!  I was stuck how to deal **kolibri-listcontent.py**.  Moving it to our kolibri-gnome fork is a good idea.

File: org.endlessos.Key.yaml [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+  - --env=KOLIBRI_HTTP_PORT=0
+  - --env=PYTHONPATH=/app/kolibri-plugins/lib/python
+
+add-extensions:
+  org.learningequality.Kolibri.Content:
+    version: '1.0'
+    directory: share/kolibri-content
+    subdirectories: true
+    no-autodownload: true
+
```
[the remark this answers is not in the document]
Comment (starnight) [reply org.endlessos.Key.yaml:25]: Ya!  That's one of the difference I spent time for testing.
I try some content channels and play the video/audio and they work.  However, I did not try all channels.

Maybe, we still take it to make sure all kinds of videos are playable.

File: org.endlessos.Key.yaml [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+sdk: org.gnome.Sdk
+command: /app/bin/kolibri-gnome
+finish-args:
+  - --device=dri
+  - --share=ipc
+  - --share=network
+  - --socket=fallback-x11
+  - --socket=pulseaudio
+  - --socket=wayland
+  - --socket=x11
```
[the remark this answers is not in the document]
Comment (starnight) [reply org.endlessos.Key.yaml:13]: sure!

File: org.endlessos.Key.yaml [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+  - modules/kolibri-content-dir.json
+  - modules/kolibri-plugins-dir.json
+
+  - name: kolibri-gnome
+    buildsystem: meson
+    builddir: true
+    run-tests: true
+    config-opts:
+      - -Dkolibri_home_template_dir=/app/share/kolibri-home-template
+      - -Dprofile=default
```
[the remark this answers is not in the document]
Comment (starnight) [reply org.endlessos.Key.yaml:59]: ok!

**Comment by starnight** [comment]
bot, build org.endlessos.Key

File: org.endlessos.Key.yaml [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+    sources:
+      - type: git
+        url: https://github.com/endlessm/endless-key-flatpak.git
+        tag: endlesskey_1.0.0
+        commit: 927b8022cc25facc11fe38ccd20ab2c04b49da1b
+
+  - name: python3-klobri-cleanup
```
[the remark this answers is not in the document]
Comment (starnight) [reply org.endlessos.Key.yaml]: Good catch!

File: org.endlessos.Key.yaml (hunk 5 above)
Comment (starnight) [reply org.endlessos.Key.yaml]: Sure!  Thanks!

File: org.endlessos.Key.yaml (hunk 6 above)
Comment (starnight) [reply org.endlessos.Key.yaml:13]: Good question!

Endless OS provides features and services from [eos-kolibri](https://github.com/endlessm/eos-kolibri) to manage the content with system resource.  Endless Key checks environment variable [`ENDLESS_KEY_USE_SYSTEM_INSTANCE`](https://github.com/endlessm/endless-key-flatpak/blob/v0.1/src/libkolibri_daemon_dbus/kolibri-daemon-dbus-utils.c#L112) before talks to the named service on the system bus.

File: org.endlessos.Key.yaml (hunk 2 above)
Comment (starnight) [reply org.endlessos.Key.yaml:83]: Have the PR [Install kolibri-listcontent.py](https://github.com/endlessm/endless-key-flatpak/pull/30)

File: org.endlessos.Key.yaml (hunk 7 above)
Comment (starnight) [reply org.endlessos.Key.yaml]: sure!

File: org.endlessos.Key.yaml (hunk 3 above)
Comment (starnight) [reply org.endlessos.Key.yaml]: Thanks!  Updated with new commits.

File: tools/kolibri-listcontent.py
Comment (starnight) [reply tools/kolibri-listcontent.py:1]: Have some discussion at https://github.com/flathub/flathub/pull/4481#discussion_r1316162995
Moving it to our repository now.

File: org.endlessos.Key.yaml (hunk 8 above)
[the remark this answers is not in the document]
Comment (starnight) [reply org.endlessos.Key.yaml:75]: If the locales should be dropped, I can prepare the command to remove them.

File: org.endlessos.Key.yaml (hunk 6 above)
Comment (hfiguiere) [reply org.endlessos.Key.yaml:13]: Given that Flathub !- Endless, how would that work on, say, Fedora.

Comment (pwithnall) [reply org.endlessos.Key.yaml:13]: aiui, on non-Endless systems, a session daemon is used instead of the system daemon. This performs the same functions, but doesn’t allow sharing of downloaded content between users.

File: org.endlessos.Key.yaml (hunk 4 above)
Comment (starnight) [reply org.endlessos.Key.yaml:61]: We have to take care the Python package installation paths and the dependencies for the combination.

**Comment by starnight** [comment]
bot, build org.endlessos.Key

**Comment by starnight** [comment]
bot, build org.endlessos.Key

**Comment by starnight** [comment]
It is good to be merged!  Thanks!

**Review by pwithnall — APPROVED** [review]

**Review by dbnicholson — APPROVED** [review]
LGTM. Nice work @starnight!

File: README.md [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+
+## Adding desktop launchers for Kolibri channels
+
+Kolibri in Endless Key generates desktop launchers as convenient shortcuts to specific channels you have installed. If you would like for these launchers to appear on your desktop, add `$HOME/.var/app/org.endlessos.Key/data/kolibri/content/xdg/share` to your `XDG_DATA_DIRS`.
+
+To achieve this for all users, create a file named `/usr/lib/systemd/user-environment-generators/61-endless-key-desktop-xdg-plugin` with the following contents, then log out and log in again:
+
+    #!/bin/bash
+    XDG_DATA_DIRS="$HOME/.var/app/org.endlessos.Key/data/kolibri/content/xdg/share:${XDG_DATA_DIRS:-/usr/local/share:/usr/share}"
+    echo "XDG_DATA_DIRS=$XDG_DATA_DIRS"
```
Comment (cassidyjames) [inline README.md:38]: Is this still relevant to Key versus upstream Kolibri?

Comment (dbnicholson) [reply README.md:38]: Yes, I think it is. Both applications provide the same feature via https://github.com/endlessm/kolibri-app-desktop-xdg-plugin.

**Comment by dbnicholson** [comment]
bot, build org.endlessos.Key

**Comment by dbnicholson** [comment]
@barthalion I think this is all ready. Can you merge it when you have a chance?
